utils/hxascdmp.1
.TH HXASCDMP "1" "May 2026" "sg3_utils\-1.49" SG3_UTILS
.SH NAME
hxascdmp \- hexadecimal ASCII dump
.SH SYNOPSIS
.B hxascdmp
[\fI\-1\fR] [\fI\-2\fR] [\fI\-b=BPL\fR] [\fI\-h\fR] [\fI\-H\fR] [\fI\-N\fR]
[\fI\-o=OFF\fR] [\fI\-q\fR] [\fI\-V\fR] [\fIFILE+\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
This utility reads one or more \fIFILE\fR names and dumps their contents
in hexadecimal and ASCII to stdout. If no \fIFILE\fR is given then stdin
is read instead; reading continues (or stalls) until an EOF is received.
The output is byte oriented, so there are no options for word (2 bytes) or
double word (4 bytes) representations.
.PP
The default format is to start each line with the hexadecimal address (offset
from the start of file) followed by 16 hexadecimal bytes separated by a
single space (apart from the 8th and 9th bytes which are separated by two
spaces; for improved readability). If the \fI\-H\fR is not given, there is
then a string of 16 ASCII characters corresponding to the hexadecimal bytes
earlier in the line; only bytes in the range 0x20 to 0x7e are printed in
ASCII, other bytes values are printed as '.' . If the \fI\-H\fR is not
given, each \fIFILE\fR name that appears on the command line is printed on
a separate line prior to that file's hexadecimal ASCII dump.
.PP
If the \fI\-N\fR option is given then no address is printed out and each
line starts with the next hexadecimal byte.
.PP
This utility is pretty close to the 'hexdump -C' variant of BSD's
.B hexdump(1)
command. And hexdump is recommended if a word or double word representation
is required.
.SH OPTIONS
.TP
\fB\-1\fR
prints the first byte of \fIFILE\fR in ASCII hexadecimal (i.e. two ASCII
characters) then exits. If there are multiple \fIFILE\fRs then the first
byte of each file is printed on a separate line. If there are no \fIFILE\fRs
then it reads the first byte from stdin, prints it in ASCII hex then exits.
To prepend '0x' to the output add the \fI\-H\fR option.
.TP
\fB\-2\fR
prints the first two bytes of \fIFILE\fR in ASCII hexadecimal (i.e. two
ASCII hex digits [0-9, a-f], a space character followed by two ASCII hex
digits) then exits. If there are multiple \fIFILE\fRs then the first
two bytes of each file is printed on a separate line. If there are no
\fIFILE\fRs then it reads the first two bytes from stdin, prints them in
ASCII hex then exits. To prepend '0x' each hex byte in the output add the
\fI\-H\fR option.
.TP
\fB\-b\fR=\fIBPL\fR
where \fIBPL\fR specifies the number of bytes per line. The default value is
16. 16 bytes per line is just enough to allow the address, 16 bytes in
hexadecimal followed by 16 bytes as ASCII to fit on a standard 80 column
wide terminal.
.TP
\fB\-h\fR
output the usage message then exit.
.TP
\fB\-H\fR
output hexadecimal only (i.e. don't place an ASCII representation at the
end of each line).
.TP
\fB\-N\fR
no address; so each line starts with the next hexadecimal byte.
.TP
\fB\-o\fR=\fIOFF\fR
where \fIOFF\fR specifies the byte offset from the beginning of the pipe or
the beginning of each file that the output starts. If the address is being
printed out then it starts at \fIOFF\fR. The default is an \fIOFF\fR of 0 .
.TP
\fB\-q\fR
Don't print any file information on the first line.
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH NOTES
In Windows the given file (or files) are set to binary mode.
.SH EXIT STATUS
The exit status of hxascdmp is 0 when it is successful. If any of the
given \fIFILE\fR names cannot be opened then the exit status is 1.
.SH EXAMPLES
First we manufacture a short file with a mix of data in it: mostly ASCII with
some control characters and 0xaa (which the echo command only accepts in
octal (0252):
.PP
   $ echo -e "three blind mice,\t\r\0252" > 3bm.txt
.PP
Now we use this utility to see exactly what is in the file. To avoid
problems with line wrapping, the bytes per line option is set to 8:
.PP
   $ hxascdmp -b=8 3bm.txt
.br
ASCII hex dump of file: 3bm.txt
.br
 00      74 68 72 65  65 20 62 6c   three bl
.br
 08      69 6e 64 20  6d 69 63 65   ind mice
.br
 10      2c 09 0d aa  0a            ,....
.PP
Using the same file, use this utility to output only hexadecimal formatted
16 bytes per line.
.PP
   $ hxascdmp -H 3bm.txt
.br
hex dump of file: 3bm.txt
.br
 00      74 68 72 65 65 20 62 6c  69 6e 64 20 6d 69 63 65
.br
 10      2c 09 0d aa 0a
.PP
For comparison the hexdump utility gives similar output:
.PP
   $ hexdump -C 3bm.txt
.br
00000000  74 68 72 65 65 20 62 6c  69 6e 64 20 6d 69 63 65  |three blind mice|
.br
00000010  2c 09 0d aa 0a                                    |,....|
.br
00000015
.PP
To paginate the output, pipe through the
.B more
or
.B less
command. That way viewing part of a long binary file can be terminated by
using 'q' when the more (or less) command pauses after displaying one
page of hex.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2004\-2026 Douglas Gilbert
.br
This software is distributed under a FreeBSD license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B hexdump(1), less(1), more(1)


ChangeLog
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for pre-release sg3_utils-1.49 [20260504] [svn: r1104]
  - AIX partial support: needs getopt.h header which
    surprisingly is non-standard. Keep hidden version in
    lib directory and copy to include dir as needed
     - still no real pass-through, using sg_dummy instead
  - getopt_long directory contents -> lib directory
  - sg_get_elem_status: fractional elements, sbc6r02
  - resync with github 20260310 after pulls accepted
  - JSON: make output more consistent so most command
    responses have a *_paramter_data or similar sub-object
  - apply https://github.com/doug-gilbert/sg3_utils/pull/39
    and its revision [20230807] mainly for Android
  - sg_inq: update version descriptors to T10 table 20230814
  - sg_wr_mode: fix --contents and --cfile= handling
    - add --mfile=MF option for decoding mask in the MF file
  - sg_vpd: inform about SPC-6 (sg_inq knew already)
    https://github.com/doug-gilbert/sg3_utils/issues/40
    - add --descriptors option, for version descriptors
  - sg_inq+sg_vpd: show depopulation time in hours and
    minutes as well as its raw form: seconds
    - add RSCS field to Block limits ext. vpage [sbc5r5]
    - translation of NVMe FR field uses sg_last_n_non_blank()
  - sg_turs: fix missing CDB on several codepaths
  - sg_requests: improve buffer truncation
  - sg_logs: correct tape Device statisitics lpage [0x14]
    serial number parameters [0x40 and 0x41] per
    https://github.com/doug-gilbert/sg3_utils/issues/45
  - sg_get_elem_status: various improvements
  - sg_get_lba_status: additional LBA accessibility values
  - sg_stream_ctl: add --hex, --inhex=, --json= and --raw
    options; additions in sbc5r05 (perm and relative_lifetime
    fields).
  - sg_dd: add grpnum=GN operand and --nocopy option
  - scsi-rescan.8 manpage added, links to rescan-scsi-bus.8
    - fix introduced Makefile.am bug, plus only for Linux
  - rescan-scsi-bus.sh: fix multipath resize without update 
    https://github.com/doug-gilbert/sg3_utils/pull/43.diff
    and 44.diff
    - fix for github.com/doug-gilbert/sg3_utils/issues/46
    - github pull requests: #68, 69 and 71 by Paul Adams
    - github pull req: #72 fix udevadm invocation M. Wilck
  - fix 'make distcheck' (again)
  - start complying with T10 23-047r1 "Broadcom SNT Reference"
    document assuming T10 SNT drafts are coming
  - sg_lib: add sg_last_n_non_blank() for SNT FR field
  - sg_lib: sg_json.c fixes from lsscsi project pull #15
  - rearrange SNT and NVMe headers and associated source
  - new error code: SG_LIB_PROGRESS_NOT_READY for long
    duration 'in progress' additional sense codes
  - lib/sg_pt_solaris.c : fix compilation issue per
    https://github.com/doug-gilbert/sg3_utils/pull/41
  - lib/sgj_haj_*() functions: tweak hex_haj argument meaning
  - scripts/00-scsi-sg3_config.rules : apply
    https://github.com/doug-gilbert/sg3_utils/pull/42
  - testing/sg_chk_inq_vd.c: test internal table against T10
    version descriptor file
  - check/clean against gcc-16 -fanalyzer; clang-22 --analyze

Changelog for released sg3_utils-1.48 [20230801] [svn: r1042]
  - decoding utilities: add --json[=JO] and --js-file=JFN
    options, short form changed to -j[=JO]
  - sg_sat_datetime: new utility for setting or accessing
    the data and time on a ATA device using a SAT layer.
  - sg_z_act_query: new utility for sending either a
    Zone activate or Zone query command
  - sg_rep_density: new utility for decoding the response of
    Report density support command [ssc (tape)]
  - sg_rem_rest_elem: new utility for removing or restoring
    elements
  - sg_write_attr: new utility, Write Attribute is a SPC
    command that is used mainly by tape drives (ssc)
  - --hex option cleanup, especially in sg_inq and sg_vpd
    - swap meaning of '-H' and '-HH' in sg_ses
  - sg_rtpg: https://github.com/hreinecke/sg3_utils/pull/79
    applied
  - netbsd: experimental port, scsi only
  - rescan-scsi-bus.sh: with '-r' it crashed due to change in
    rev 867 in sg_inq.c from Device_type= to PDT= .
    Change script to use either
    - undo regression in rev 815 that added newline after
      each LUN in the default (no option) output
    - rev 815 changed the order of listing hosts from
      numeric to alphabetical, change it back to numeric
    - https://github.com/doug-gilbert/sg3_utils/pull/17
      applied with tweaks: add timeout parameter
    - clean $norm handling
    - fix handling of '-I <secs>' option
    - sgdevice26: do not traverse sg class if scsi_device
      is not added
    - add -no-lip-scan option
    - https://github.com/doug-gilbert/sg3_utils/pull/21
    - speed testonline() when peripheral_qualifier != 0
      see https://github.com/doug-gilbert/sg3_utils/issues/24
    - improve handling of USB SD card readers (with no cards
      present)
    - speed multipath scans with many LUNs, cache multipath
      LUN info in temporary file, see
      https://github.com/doug-gilbert/sg3_utils/issues/22
  - udev rules: restrict use of ambiguous device IDs patchset
    (3 part) from Martin Wilck applied
  - sg_rep_zones: add Report zone starting LBA granularity
    field in REPORT ZONES response [zbc2r12]
    - add --brief option, show part of header and last
      descriptor fetched
    - add --find ZT option to find the first occurrence of
      ZT; if ZT prefixed by - or ! find first not equal to ZT
    - add --statistics option
  - sg_get_elem_status: change '--maxlen=' option default to
    1056 (header plus 32 physical element status descriptors)
    - add current (maximum) number of depopulated elements
      fields
  - sg_get_lba_status: decode new LBA accessibility field
  - sg_decode_sense: add --nodecode option
  - sg_logs: tweak the meaning of --list option to more closely

utils/hxascdmp.c
/*
 * Copyright (c) 2004-2026 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#define DEF_BYTES_PER_LINE 16

static int bytes_per_line = DEF_BYTES_PER_LINE;

static const char * version_str = "1.12 20260504";

#define CHARS_PER_HEX_BYTE 3
#define BINARY_START_COL 6
#define MAX_LINE_LENGTH 257


#ifdef SG_LIB_MINGW
/* Non Unix OSes distinguish between text and binary files.
   Set text mode on fd. Does nothing in Unix. Returns negative number on
   failure. */
int
sg_set_text_mode(int fd)
{
    return setmode(fd, O_TEXT);
}

/* Set binary mode on fd. Does nothing in Unix. Returns negative number on
   failure. */
int
sg_set_binary_mode(int fd)
{
    return setmode(fd, O_BINARY);
}

#else
/* For Unix the following functions are dummies. */
int
sg_set_text_mode(int fd)
{
    return fd;  /* fd should be >= 0 */
}

int
sg_set_binary_mode(int fd)
{
    return fd;
}
#endif

/* Returns the number of times 'ch' is found in string 's' given the
 * string's length. */
static int
num_chs_in_str(const char * s, int slen, int ch)
{
    int res = 0;

    while (--slen >= 0) {
        if (ch == s[slen])
            ++res;
    }
    return res;
}

/* If the number in 'buf' can be decoded or the multiplier is unknown
 * then -1LL is returned. Accepts a hex prefix (0x or 0X) or a decimal
 * multiplier suffix (as per GNU's dd (since 2002: SI and IEC 60027-2)).
 * Main (SI) multipliers supported: K, M, G, T, P. Ignore leading spaces
 * and tabs; accept comma, hyphen, space, tab and hash as terminator. */
int64_t
sg_get_llnum(const char * buf)
{
    int res, len, n;
    int64_t num, ll;
    uint64_t unum;
    char * cp;
    const char * b;
    char c = 'c';
    char c2 = '\0';     /* keep static checker happy */
    char c3 = '\0';     /* keep static checker happy */
    char lb[32];

    if ((NULL == buf) || ('\0' == buf[0]))
        return -1LL;
    len = strlen(buf);
    n = strspn(buf, " \t");
    if (n > 0) {
        if (n == len)
            return -1LL;
        buf += n;
        len -= n;
    }
    /* following hack to keep C++ happy */
    cp = strpbrk((char *)buf, " \t,#-");
    if (cp) {
        len = cp - buf;
        n = (int)sizeof(lb) - 1;
        len = (len < n) ? len : n;
        memcpy(lb, buf, len);
        lb[len] = '\0';
        b = lb;
    } else
        b = buf;
    if (('0' == b[0]) && (('x' == b[1]) || ('X' == b[1]))) {
        res = sscanf(b + 2, "%" SCNx64 , &unum);
        num = unum;
    } else if ('H' == toupper((int)b[len - 1])) {
        res = sscanf(b, "%" SCNx64 , &unum);
        num = unum;
    } else
        res = sscanf(b, "%" SCNd64 "%c%c%c", &num, &c, &c2, &c3);
    if (res < 1)
        return -1LL;
    else if (1 == res)
        return num;
    else {
        if (res > 2)
            c2 = toupper((int)c2);
        if (res > 3)
            c3 = toupper((int)c3);
        switch (toupper((int)c)) {
        case 'C':
            return num;
        case 'W':
            return num * 2;
        case 'B':
            return num * 512;
        case 'K':
            if (2 == res)
                return num * 1024;
            if (('B' == c2) || ('D' == c2))
                return num * 1000;
            if (('I' == c2) && (4 == res) && ('B' == c3))
                return num * 1024;
            return -1LL;
        case 'M':
            if (2 == res)
                return num * 1048576;
            if (('B' == c2) || ('D' == c2))
                return num * 1000000;
            if (('I' == c2) && (4 == res) && ('B' == c3))
                return num * 1048576;
            return -1LL;
        case 'G':
            if (2 == res)
                return num * 1073741824;
            if (('B' == c2) || ('D' == c2))
                return num * 1000000000;
            if (('I' == c2) && (4 == res) && ('B' == c3))
                return num * 1073741824;
            return -1LL;
        case 'T':
            if (2 == res)
                return num * 1099511627776LL;
            if (('B' == c2) || ('D' == c2))
                return num * 1000000000000LL;
            if (('I' == c2) && (4 == res) && ('B' == c3))
                return num * 1099511627776LL;
            return -1LL;
        case 'P':
            if (2 == res)
                return num * 1099511627776LL * 1024;
            if (('B' == c2) || ('D' == c2))
                return num * 1000000000000LL * 1000;
            if (('I' == c2) && (4 == res) && ('B' == c3))
                return num * 1099511627776LL * 1024;
            return -1LL;
        case 'X':
            cp = (char *)strchr(b, 'x');
            if (NULL == cp)
                cp = (char *)strchr(b, 'X');
            if (cp) {
                ll = sg_get_llnum(cp + 1);
                if (-1LL != ll)
                    return num * ll;
            }
            return -1LL;
        default:
            fprintf(stderr, "unrecognized multiplier\n");
            return -1LL;
        }
    }
}

static void
dStrHex(const char* str, int len, long start, int noAddr)
{
    const char* p = str;
    unsigned char c;
    char buff[MAX_LINE_LENGTH];
    long a = start;
    int bpstart, cpstart;
    int j, k, line_length, nl, cpos, bpos, midline_space;

    if (noAddr) {
        bpstart = 0;
        cpstart = ((CHARS_PER_HEX_BYTE * bytes_per_line) + 1) + 5;
    } else {
        bpstart = BINARY_START_COL;
        cpstart = BINARY_START_COL +
                        ((CHARS_PER_HEX_BYTE * bytes_per_line) + 1) + 5;
    }
    cpos = cpstart;
    bpos = bpstart;
    midline_space = ((bytes_per_line + 1) / 2);

    if (len <= 0)
        return;
    line_length = BINARY_START_COL +
                  (bytes_per_line * (1 + CHARS_PER_HEX_BYTE)) + 7;
    if (line_length >= MAX_LINE_LENGTH) {
        fprintf(stderr, "bytes_per_line causes maximum line length of %d "
                        "to be exceeded\n", MAX_LINE_LENGTH);
        return;
    }
    memset(buff, ' ', line_length);
    buff[line_length] = '\0';
    if (0 == noAddr) {
        k = sprintf(buff + 1, "%.2lx", a);
        buff[k + 1] = ' ';
    }

    for(j = 0; j < len; j++) {
        nl = (0 == (j % bytes_per_line));
        if ((j > 0) && nl) {
            printf("%s\n", buff);
            bpos = bpstart;
            cpos = cpstart;
            a += bytes_per_line;
            memset(buff,' ', line_length);
            if (0 == noAddr) {
                k = sprintf(buff + 1, "%.2lx", a);
                buff[k + 1] = ' ';
            }
        }
        c = *p++;
        bpos += (nl && noAddr) ?  0 : CHARS_PER_HEX_BYTE;
        if ((bytes_per_line > 4) && ((j % bytes_per_line) == midline_space))
            bpos++;
        sprintf(&buff[bpos], "%.2x", (int)(unsigned char)c);
        buff[bpos + 2] = ' ';
        if ((c < ' ') || (c >= 0x7f))
            c='.';
        buff[cpos++] = c;
    }
    if (cpos > cpstart)
        printf("%s\n", buff);
}

static void
dStrHexOnly(const char* str, int len, long start, int noAddr)
{
    const char* p = str;
    unsigned char c;
    char buff[MAX_LINE_LENGTH];
    long a = start;
    int bpstart, bpos, nl;
    int midline_space = ((bytes_per_line + 1) / 2);
    int j, k, line_length;

    if (len <= 0)
        return;
    bpstart = (noAddr ? 0 : BINARY_START_COL);
    bpos = bpstart;
    line_length = (noAddr ? 0 : BINARY_START_COL) +
                  (bytes_per_line * CHARS_PER_HEX_BYTE) + 4;
    if (line_length >= MAX_LINE_LENGTH) {
        fprintf(stderr, "bytes_per_line causes maximum line length of %d "
                        "to be exceeded\n", MAX_LINE_LENGTH);
        return;
    }
    memset(buff, ' ', line_length);
    buff[line_length] = '\0';
    if (0 == noAddr) {
        k = sprintf(buff + 1, "%.2lx", a);
        buff[k + 1] = ' ';
    }

    for(j = 0; j < len; j++) {
        nl = (0 == (j % bytes_per_line));
        if ((j > 0) && nl) {
            printf("%s\n", buff);
            bpos = bpstart;
            a += bytes_per_line;
            memset(buff,' ', line_length);
            if (0 == noAddr) {
                k = sprintf(buff + 1, "%.2lx", a);
                buff[k + 1] = ' ';
            }
        }
        c = *p++;
        bpos += (nl && noAddr) ? 0 : CHARS_PER_HEX_BYTE;
        if ((bytes_per_line > 4) && ((j % bytes_per_line) == midline_space))
            bpos++;
        sprintf(&buff[bpos], "%.2x", (int)(unsigned char)c);
        buff[bpos + 2] = ' ';
    }
    if (bpos > bpstart)
        printf("%s\n", buff);
}

static void
usage()
{
    fprintf(stderr, "Usage: hxascdmp [-1] [-2] [-b=<n>] [-h] [-H] [-N] "
            "[-o=<off>] [-q]\n"
            "                [-V] [-?]  [<file>+]\n");
    fprintf(stderr, "  where:\n");
    fprintf(stderr, "    -1         print first byte in hex, prepend '0x' "
            "if '-H' given\n");
    fprintf(stderr, "    -2         like '-1' but print first two bytes\n");
    fprintf(stderr, "    -b=<n>     bytes per line to display "
                    "(def: 16)\n");
    fprintf(stderr, "    -h         print this usage message\n");
    fprintf(stderr, "    -H         print hex only (i.e. no ASCII "
            "to right)\n");
    fprintf(stderr, "    -N         no address, start in first column\n");
    fprintf(stderr, "    -o=<off>    start decoding at byte <off>. Suffix "
            "multipliers allowed\n");
    fprintf(stderr, "    -q         quiet: suppress output of header "
            "info\n");
    fprintf(stderr, "    -V         print version string then exits\n");
    fprintf(stderr, "    -?         print this usage message\n");
    fprintf(stderr, "    <file>+    reads file(s) and outputs each "
                    "as hex ASCII\n");
    fprintf(stderr, "               if no <file> then reads stdin\n\n");
    fprintf(stderr, "Sends hex ASCII dump of stdin/file to stdout\n");
}

int
main(int argc, const char ** argv)
{
    char buff[8192];
    int num = 8192;
    long start = 0;
    int64_t offset = 0;
    int res, k, u, len, n;
    int inFile = STDIN_FILENO;
    int doHelp = 0;
    int doHex = 0;
    int noAddr = 0;
    int doVersion = 0;
    int hasFilename = 0;
    int quiet = 0;
    int print1 = 0;
    int print2 = 0;
    int ret = 0;
    const char * cp;

    for (k = 1; k < argc; k++) {
        cp = argv[k];
        len = strlen(cp);
        if (0 == strncmp("-b=", cp, 3)) {
            res = sscanf(cp + 3, "%d", &u);
            if ((1 != res) || (u < 1)) {
                fprintf(stderr, "Bad value after '-b=' option\n");
                usage();
                return 1;
            }
            bytes_per_line = u;
        } else if (0 == strncmp("-o=", cp, 3)) {
            int64_t off = sg_get_llnum(cp + 3);

            if (off == -1) {
                fprintf(stderr, "Bad value after '-o=' option\n");
                usage();
                return 1;
            }
            offset = off;
        } else if ((len > 1) && ('-' == cp[0]) && ('-' != cp[1])) {
            res = 0;
            n = num_chs_in_str(cp + 1, len - 1, '1');
            print1 += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, '2');
            print2 += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, 'h');
            doHelp += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, 'H');
            doHex += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, 'N');
            noAddr += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, 'q');
            quiet += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, 'V');
            doVersion += n;
            res += n;
            n = num_chs_in_str(cp + 1, len - 1, '?');
            doHelp += n;
            res += n;
            if (0 == res) {
                fprintf(stderr, "No option recognized in str: %s\n", cp);
                usage();
                return 1;
            }
        } else if (0 == strcmp("-?", argv[k]))
            ++doHelp;
        else if (*argv[k] == '-') {
            fprintf(stderr, "unknown switch: %s\n", argv[k]);
            usage();
            return 1;
        } else {
            hasFilename = 1;
            break;
        }
        if (print2)
            print1 += print2 + print2;
    }
    if (doVersion) {
        printf("%s\n", version_str);
        return 0;
    }
    if (doHelp) {
        usage();
        return 0;
    }

    /* Make sure num to fetch is integral multiple of bytes_per_line */
    if (0 != (num % bytes_per_line))
        num = (num / bytes_per_line) * bytes_per_line;

    if (hasFilename) {
        for ( ; k < argc; k++)
        {
            inFile = open(argv[k], O_RDONLY);
            if (inFile < 0) {
                fprintf(stderr, "Couldn't open file: %s\n", argv[k]);
                ret = 1;
            } else {
                sg_set_binary_mode(inFile);
                if (offset > 0) {
                    int err;
                    int64_t off_res;

                    off_res = lseek(inFile, offset, SEEK_SET);
                    if (off_res < 0) {
                        err = errno;
                        fprintf(stderr, "failed moving filepos: wanted=%"
                                PRId64 " [0x%" PRIx64 "]\nlseek error: %s\n",
                                offset, offset, strerror(err));
                        goto fini1;
                    }
                    start = offset;
                } else
                    start = 0;
                if (! (doHex || quiet || print1))
                    printf("ASCII hex dump of file: %s\n", argv[k]);
                while ((res = read(inFile, buff, num)) > 0) {
                    if (print1) {
                        if (1 == print1) {
                            if (doHex)
                                printf("0x%02x\n", (uint8_t)(buff[0]));
                            else
                                printf("%02x\n", (uint8_t)(buff[0]));
                        } else {
                            uint16_t us;

                            memcpy(&us, buff, 2);
                            if (doHex)
                                printf("0x%04x\n", us);
                            else
                                printf("%04x\n", us);
                        }
                        break;
                    }
                    if (doHex)
                        dStrHexOnly(buff, res, start, noAddr);
                    else
                        dStrHex(buff, res, start, noAddr);
                    start += (long)res;
                }
            }
fini1:
            close(inFile);
        }
    } else {
        sg_set_binary_mode(inFile);
        if (offset > 0) {
            start = offset;
            do {        /* eat up offset bytes */
                if ((res = read(inFile, buff,
                                (num > offset ? offset : num))) > 0)
                    offset -= res;
                else {
                    fprintf(stderr, "offset read() error: %s\n",
                            strerror(errno));
                     break;
                }
            } while (offset > 0);
        }
        while ((res = read(inFile, buff, num)) > 0) {
            if (doHex)
                dStrHexOnly(buff, res, start, noAddr);
            else
                dStrHex(buff, res, start, noAddr);
            start += (long)res;
        }
    }
    return ret;
}
